### ISSUE TITLE
Condition on Trainable = False with shared embedding layers

### ISSUE THREAD

**Issue by edmondja** [keras-team/keras#8259]
I am using an embedding layer which I use in two locations of my network as input, as a shared layer.
**And I want conditionally to where this shared embedding layer is used in the network to make it not trainable.**

To make it simple lets put the problem this way, I have two neural networks 1 and 2 which takes as inputs the same embedding layer and which are regrouped with a "concatenate" layer at the end of the network.
I want back-propagation to stop in the network 2 just before it passes through the embedding layer (again).

I tried to add a dropout layer with p = 0, but not only I discovered that it didn't remplace the functionality "Trainable = False", but I was also very surprised to see that you can build a neural network putting everywhere dropout with p = 0 and it still trains ! **How is that possible ?**

So I actually have two questions.

**Comment by ablacklama** [comment]
I'm new to this but if I understand your question, you are wondering how adding dropout layer (model.add(Dropout(0.0)) throughout your network doesn't freeze the weights for training?

Dropout (as you probably know) is the process of setting random inputs to 0 whenever the network is run. The value passed to the Keras dropout layer is the percent of the data to set to 0. This means that passing it '0' as an argument will dropout none of the inputs leaving the data unchanged. This doesn't effect trainability at all and acts just like any other keras layer during the training process. 
If you're talking about dropping out all the data and trying to train it, that's a different story. If you set all the inputs to 0 during the training process then it won't train. If it is still training it means that there wasn't a 100% dropout rate on the last layer at least. If a dense layer is your output for instance and there is no dropout layer after it, the network will just try to optimize the bias of the last layer to produce the best output it can find.

Hope this helps!

**Comment by edmondja** [comment]
@ablacklama Thank you but even when I put p=1 it still fits.
ie. : if I set dropout of a dense layer to p=1 and I print : 
model.layers[0].get_weights()
the value changes ! How is that possible ?

**Comment by edmondja** [comment]
Nobody ? Its a real issue I am facing here.

**Comment by dksakkos** [comment]
Hi @edmondja, I was wondering if you found a way to solve this problem?

**Comment by edmondja** [comment]
Hello, no, that’s why I didn’t close the thread.

**Comment by edmondja** [comment]
> Hi @edmondja, I was wondering if you found a way to solve this problem?

I think somebody (maybe me) should develop a new custom callback which copies the weights of a layer to another layer at each training batch.

This way you can make a similar layer but frozen and copy after each training batch the weights of the non frozen one.
